src/WebJobs.Extensions.DurableTask/Listener/FunctionExecutionHelper.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Listener
+{
+    internal static class FunctionExecutionHelper
+    {
+        public static async Task<WrappedFunctionResult> ExecuteFunctionInOrchestrationMiddleware(
+            ITriggeredFunctionExecutor executor,
+            IApplicationLifetimeWrapper hostServiceLifetime,
+            TriggeredFunctionData triggerInput,
+            CancellationToken cancellationToken)
+        {
+#pragma warning disable CS0618 // InvokeHandler approved for use by this extension
+            if (triggerInput.InvokeHandler == null)
+            {
+                throw new ArgumentException(
+                    $"{nameof(ExecuteFunctionInOrchestrationMiddleware)} should only be used when ${nameof(triggerInput)} has a value for ${nameof(TriggeredFunctionData.InvokeHandler)}");
+            }
+
+            try
+            {
+                bool executedUserCode = false;
+                var triggeredFunctionData = new TriggeredFunctionData()
+                {
+                    TriggerValue = triggerInput.TriggerValue,
+                    ParentId = triggerInput.ParentId,
+                    InvokeHandler = async userCodeHandler =>
+                    {
+                        executedUserCode = true;
+                        await triggerInput.InvokeHandler(userCodeHandler);
```

Comment:
nit: I suggest removing `async` and `await` here to reduce the number of exceptions that get bubbled up in the failure code path.